### ISSUE TITLE
Alsa demo not recognizing anything 

### ISSUE THREAD

**Issue by danicalifornia** [Picovoice/Porcupine#104]
### Expected behavior
I compiled the C++ single keyword alsa demo, set pcm card with usb mic as a default and was able to start also demo with no errors. Also demo displays „Listen...“ and should now detect spoken key words.

### Actual behavior 

Alsa demo does not detect spoken key words at all. Just states „Listen...“ but does not react at all to keywords spoken. I selected different keyword files and re-compiled - same issue...

### Steps to reproduce the behavior

Just repeat the steps described above

**Comment by kenarsa** [comment]
@oziee could you please help out and see if you can address this issue?

**Comment by oziee** [comment]
hey @danicalifornia 

If you unplug the USB device does the ALSA Demo app still run saying "listening.." or does it throw an error??

if you type in `arecord -l` does it list the USB mic? if no device is listed then it can not see the mic.. if the mic is listed then you need to made sure the /etc/asound.conf is configured to use it 

we can test to see if the system can hear the mic
`arecord -f cd test.wav`
then listen back 
`aplay test.wav`

the code is set to "default" device but this can be altered in the code itself or from the `/etc/asound.conf` file which Alsa uses 
`
arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: Device [USB PnP Sound Device], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
`
the device is hw:1,0 (card 1 device 0)
so the /etc/asound.conf file needs to look something list this: (your device and card numbers for input/ouput might be different)
`
pcm.!default {
  type asym
   playback.pcm {
     type plug
     slave.pcm "hw:0,0"
   }
   capture.pcm {
     type plug
     slave.pcm "hw:1,0"
   }
}
`

**Comment by danicalifornia** [comment]
Hi @oziee,

thank you very much for your detailed resonse!

I tried it out - unplugging the USB mic does not yield in an error.

capturing sound via arecord and playing the recorded sound file via aplay works - I had tried that already prior to posting the issue here.

Changing from "default" sound device to "hw:1,0" (which is the correct sound device named in ~/.asoundrc did not change anything unfortunately (even though if I change the value to a sound device not existing like "test", then I get an error message). alsodemo still says "Listen" but does not react no matter if sound device is set to "default" or "hw:1,0" (but that should also be the case as sound device "hw:1,0" is set do be the default sound recording device in system settings). More and more I get the feeling my mic could be not have the right sample rate or something like that - could that be the reason?

Thank you so much and best regards
danicalifornia

**Comment by oziee** [comment]
true.. a lot of USB mics are 48k these days.. and porcupine uses 16k

you can create a test wave using the `arecord` and then do a properties on it to see what the default the mic uses

you can set the rate in the  `.asoundrc` adding to the capture settings 
```
rate 16000
```
